test/e2e/helm_test.go

```diff
@@ -217,6 +217,118 @@ var _ = Describe("Workload cluster creation", func() {
+					WaitForControlPlaneInitialized: EnsureControlPlaneInitialized,
+				}),
+			), result)
+
+			result2 := new(clusterctl.ApplyClusterTemplateAndWaitResult)
+
+			clusterName2 := fmt.Sprintf("%s-%s", specName, util.RandomString(6))
+			clusterctl.ApplyClusterTemplateAndWait(ctx, createApplyClusterTemplateInput(
+				specName,
+				withNamespace(namespace.Name),
+				withClusterName(clusterName2),
+				withControlPlaneMachineCount(1),
+				withWorkerMachineCount(1),
+				withControlPlaneWaiters(clusterctl.ControlPlaneWaiters{
+					WaitForControlPlaneInitialized: EnsureControlPlaneInitialized,
+				}),
+			), result2)
+
+			hcp := &addonsv1alpha1.HelmChartProxy{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "nginx-ingress",
+					Namespace: namespace.Name,
+				},
+				Spec: addonsv1alpha1.HelmChartProxySpec{
+					ClusterSelector: metav1.LabelSelector{
+						MatchLabels: map[string]string{
+							"nginxIngress": "enabled",
+						},
+					},
+					ReleaseName:      "nginx-ingress",
+					ReleaseNamespace: "nginx-namespace",
+					ChartName:        "nginx-ingress",
+					RepoURL:          "https://helm.nginx.com/stable",
+					ValuesTemplate:   nginxValues,
+				},
+			}
+
+			// Create anew HelmChartProxy and install on Cluster 1
+			By("Creating new HelmChartProxy to install nginx on Cluster 1", func() {
+				HelmInstallSpec(ctx, func() HelmInstallInput {
+					return HelmInstallInput{
+						BootstrapClusterProxy: bootstrapClusterProxy,
+						Namespace:             namespace,
+						ClusterName:           clusterName,
+						HelmChartProxy:        hcp,
+					}
+				})
+			})
+
+			// Patch Cluster 2 labels to match HelmChartProxy's clusterSelector.
+			By("Patching Cluster 2 labels to install nginx", func() {
+				installInput := &HelmInstallInput{
+					BootstrapClusterProxy: bootstrapClusterProxy,
+					Namespace:             namespace,
+					ClusterName:           clusterName2,
+					HelmChartProxy:        hcp,
+				}
+				EnsureHelmReleaseInstallOrUpgrade(ctx, specName, bootstrapClusterProxy, installInput, nil, true)
+			})
+
+			// Uninstall Helm chart from Cluster 1 by removing the label selector.
+			By("Uninstalling Helm chart from Cluster 1", func() {
+				HelmUninstallSpec(ctx, func() HelmUninstallInput {
+					return HelmUninstallInput{
+						BootstrapClusterProxy: bootstrapClusterProxy,
+						Namespace:             namespace,
+						ClusterName:           clusterName,
+						HelmChartProxy:        hcp,
+					}
+				})
+			})
+
+			// Ensure that Helm chart is still installed on Cluster 2.
+			By("Ensuring that nginx is still installed on Cluster 2", func() {
+				installInput := &HelmInstallInput{
+					BootstrapClusterProxy: bootstrapClusterProxy,
+					Namespace:             namespace,
+					ClusterName:           clusterName2,
+					HelmChartProxy:        hcp,
+				}
+				EnsureHelmReleaseInstallOrUpgrade(ctx, specName, bootstrapClusterProxy, installInput, nil, false)
+			})
+
+			defer func() {
```

Comment:
Should this `defer func()` be higher up, like line 249 right after Cluster 2 is created? Or maybe you can put it in a [`DeferCleanup`](https://onsi.github.io/ginkgo/#cleaning-up-our-cleanup-code-defercleanup) at this level Otherwise it seems like there's a possbility it won't be cleaned up if an error happens in part of this spec.

Reply:
Moved the defer call up, let me know if this looks fine now.